### PR TITLE
Fix multinode tests script

### LINKED ISSUE #1627: run_multinode_tests_in_docker.sh :FileNotFoundError: [Errno 2] No such file or directory: 'python3.6'
## 🐛 Bug description

When running `bash ./test/run_multinode_tests_in_docker.sh`
```
FileNotFoundError: [Errno 2] No such file or directory: 'python3.6' 
```
This script is not in the CI but is useful to run the distributed tests on a local machine.
Idea would be to make it compatible with `pytorch/pytorch:latest`, eventually make an inline build of an image with pre-installed tests execution environment, add `nnodes`, `nproc_per_node`, `gpu` as script arguments.

## Environment

- PyTorch Version (e.g., 1.4):
- Ignite Version (e.g., 0.3.0):
- OS (e.g., Linux):
- How you installed Ignite (`conda`, `pip`, source):
- Python version:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by fco-dv**
Fixes #1627

Description: Try to fix the run_miltinode_tests_in_docker.sh 
- Add scripts parameters `nnodes` | `nproc_per_node` | `gpu`
- Enable gpu tests
- Put the `docker rm` steps at the end 
- Tests results on multiple configurations  CPU and GPU(s) : GPUs tests are almost OK but should be investigated because the behavior is not identical when some tests are previsously deactivated ( see below the different configurations) ( thanks @sdesrozis for your help ! )

```
Run on a node with 2 GPUs

 

Script test configurations :

nnodes |  nproc_per_nodes | gpu(enabled)

 

2  |  1   |  0

 

=============================== warnings summary ===============================

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

  /workspace/ignite/metrics/precision.py:29: RuntimeWarning: Precision/Recall metrics do not work in distributed setting when average=False and is_multilabel=True. Results are not reduced across computing devices. Computed result corresponds to the local rank's (single process) result.

    RuntimeWarning,

 

-- Docs: https://docs.pytest.org/en/stable/warnings.html

 

================= 24 passed, 26 skipped, 2 warnings in 32.53s ==================

 

2  |  4   |  0   ( default)

 

=============================== warnings summary ===============================

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

  /workspace/ignite/metrics/precision.py:29: RuntimeWarning: Precision/Recall metrics do not work in distributed setting when average=False and is_multilabel=True. Results are not reduced across computing devices. Computed result corresponds to the local rank's (single process) result.

    RuntimeWarning,

 

-- Docs: https://docs.pytest.org/en/stable/warnings.html

=========================== short test summary info ============================

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

FAILED tests/ignite/metrics/test_loss.py::test_multinode_distrib_cpu - Assert...

FAILED tests/ignite/metrics/test_loss.py::test_multinode_distrib_cpu - Assert...

FAILED tests/ignite/metrics/test_loss.py::test_multinode_distrib_cpu - Assert...

FAILED tests/ignite/metrics/test_loss.py::test_multinode_distrib_cpu - Assert...

 

======= 8 failed, 88 passed, 98 skipped, 8 warnings in 441.54s (0:07:21) =======

 

 

4  |  4  |  0

 

=============================== warnings summary ===============================

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu

  /workspace/ignite/metrics/precision.py:29: RuntimeWarning: Precision/Recall metrics do not work in distributed setting when average=False and is_multilabel=True. Results are not reduced across computing devices. Computed result corresponds to the local rank's (single process) result.

    RuntimeWarning,

 

-- Docs: https://docs.pytest.org/en/stable/warnings.html

=========================== short test summary info ============================

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

FAILED tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu

 

======= 4 failed, 92 passed, 98 skipped, 8 warnings in 654.51s (0:10:54) =======

 

 

GPU configuration ( 1 node < - > 1 gpu ):

 

2  |  1   | 1

 

After the first test passed  tests/ignite/contrib/engines/test_common.py::test_multinode_distrib_gpu , all the remaining tests failed :

 

================== 1 passed, 26 skipped, 23 errors in 45.69s ===================

 

        if _default_pg is not None:

>           raise RuntimeError("trying to initialize the default process group "

                               "twice!")

E           RuntimeError: trying to initialize the default process group twice!

 

/opt/conda/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:387: RuntimeError

 

 

When ignoring some these tests with : -k ‘not test_common’ , the remaining tests are OK

 

=============================== warnings summary ===============================

tests/ignite/metrics/test_precision.py::test_multinode_distrib_gpu

tests/ignite/metrics/test_recall.py::test_multinode_distrib_gpu

  /workspace/ignite/metrics/precision.py:29: RuntimeWarning: Precision/Recall metrics do not work in distributed setting when average=False and is_multilabel=True. Results are not reduced across computing devices. Computed result corresponds to the local rank's (single process) result.

    RuntimeWarning,

 

-- Docs: https://docs.pytest.org/en/stable/warnings.html

 

================= 23 passed, 25 skipped, 2 warnings in 38.83s ==================

```

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @fco-dv ! Looks good to me 👍 
Let me also try it on my infra.

Do you think we should integrate to Circle CI ?

**Comment by vfdev-5** [comment]
As for failing tests, let decorate `tests/ignite/engine/test_deterministic.py::test_multinode_distrib_cpu` as we allow it to fail.

What are the problems with `FAILED tests/ignite/metrics/test_loss.py::test_multinode_distrib_cpu` ?

**Comment by fco-dv** [comment]
for ` FAILED tests/ignite/metrics/test_loss.py::test_multinode_distrib_cpu` , it's a float precision error, here are the details :
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/ignite/metrics/test_loss.py:109: in _test_distrib_compute_on_criterion
    _test("cpu")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>       assert_almost_equal(res, true_loss_value.item())
E       AssertionError: 
E       Arrays are not almost equal to 7 decimals
E        ACTUAL: 1.1080787181854248
E        DESIRED: 1.108078956604004

tests/ignite/metrics/test_loss.py:107: AssertionError
```

**Comment by sdesrozis** [comment]
@fco-dv have you tackled the issue concerning `test_common` that we have been facing recently ?

**Comment by fco-dv** [comment]
@sdesrozis not yet , my guess is  that the process group is not destroyed at the end of `test_common` training step, but not quite sure how to address that...

**Comment by vfdev-5** [comment]
@fco-dv what's the issue it is about ?

**Comment by fco-dv** [comment]
@vfdev-5  when running with gpu enabled, `test_common`  PASSED but others FAILED due to `RuntimeError: trying to initialize the default process group twice!` , but when skipping this test, others PASSED.

**Comment by vfdev-5** [comment]
> @vfdev-5 when running with gpu enabled, `test_common` PASSED but others FAILED due to `RuntimeError: trying to initialize the default process group twice!` , but when skipping this test, others PASSED.

Maybe, we can try to do same as here : https://github.com/pytorch/ignite/blob/master/tests/ignite/conftest.py#L104 ?

**Comment by fco-dv** [comment]
Managed to fix gpu tests with :
https://github.com/fco-dv/ignite/commit/697dfaed5a9313cefcdf100245eb1c7614a9a190#diff-939e5dec7f6dffa9b3cd5721a69781e2cbc81b052c5676d8a1a0667915f808f5
There is certainly a more elegant way to do this ... 
test configuration : 
2 | 1 | 1  

```
=============================== warnings summary ===============================
tests/ignite/metrics/test_precision.py::test_multinode_distrib_gpu
tests/ignite/metrics/test_recall.py::test_multinode_distrib_gpu
  /workspace/ignite/metrics/precision.py:29: RuntimeWarning: Precision/Recall metrics do not work in distributed setting when average=False and is_multilabel=True. Results are not reduced across computing devices. Computed result corresponds to the local rank's (single process) result.
    RuntimeWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================= 24 passed, 26 skipped, 2 warnings in 37.98s ==================
```

**Comment by fco-dv** [comment]
Seems ok now  for: 

Default conf : 2 | 4 | 0
```
=============================== warnings summary ===============================
tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu
tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu
tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu
tests/ignite/metrics/test_precision.py::test_multinode_distrib_cpu
tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu
tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu
tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu
tests/ignite/metrics/test_recall.py::test_multinode_distrib_cpu
  /workspace/ignite/metrics/precision.py:29: RuntimeWarning: Precision/Recall metrics do not work in distributed setting when average=False and is_multilabel=True. Results are not reduced across computing devices. Computed result corresponds to the local rank's (single process) result.
    RuntimeWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
====== 92 passed, 98 skipped, 4 xfailed, 8 warnings in 507.07s (0:08:27) =======
```
and with gpu : 2 | 1 | 1 
```
=============================== warnings summary ===============================
tests/ignite/metrics/test_precision.py::test_multinode_distrib_gpu
tests/ignite/metrics/test_recall.py::test_multinode_distrib_gpu
  /workspace/ignite/metrics/precision.py:29: RuntimeWarning: Precision/Recall metrics do not work in distributed setting when average=False and is_multilabel=True. Results are not reduced across computing devices. Computed result corresponds to the local rank's (single process) result.
    RuntimeWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============ 24 passed, 26 skipped, 2 warnings in 77.77s (0:01:17) =============
```

@vfdev-5 for the CI integration would you like me to create another PR or continue on this one ? thanks!

**Comment by vfdev-5** [comment]
@fco-dv Thanks ! Let's merge it like that and for Circle CI, I'll enable it on PRs and let's intergrate it in another PR.

**Review by vfdev-5 — APPROVED** [review]

File: tests/ignite/metrics/test_loss.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def test_reset():
         loss.compute()
 
 
-def _test_distrib_compute_on_criterion(device):
+def _test_distrib_compute_on_criterion(device, tol=1e-6):
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_loss.py:78]: Actually, here we would like to use reduced tolerance only for multinode case and keep default tolerance for other cases.

**Review by vfdev-5 — CHANGES_REQUESTED** [review]

File: tests/ignite/metrics/test_loss.py (hunk 1 above)
Comment (fco-dv) [reply tests/ignite/metrics/test_loss.py]: Yes 've just made a fix for this, thanks for your comment

**Review by vfdev-5 — APPROVED** [review]
Thanks for the PR @fco-dv !

**Comment by vfdev-5** [comment]
@fco-dv tests are failing : https://github.com/pytorch/ignite/pull/1631/checks?check_run_id=1898395835

**Pull request merged**